File: app/src/main/java/ru/filchacov/billsplittest/AuthMVP/AuthPresenter.java
package ru.filchacov.billsplittest.AuthMVP;

import android.net.Uri;
import android.provider.ContactsContract;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Objects;

import ru.filchacov.billsplittest.App;
import ru.filchacov.billsplittest.ModelDB;
import ru.filchacov.billsplittest.db.User;
import ru.filchacov.billsplittest.db.UserDB;
import ru.filchacov.billsplittest.db.UserDao;

class AuthPresenter {
    private ModelDB model = new ModelDB();
    private AuthInterface view;
    private UserDB userDB = App.getInstance().getDatabase();
    private UserDao userDao = userDB.getuserDao();
    private User currentUser;

    AuthPresenter(AuthInterface view) {
        this.view = view;
    }

    void init() {
        model.getAuthReference();
        if (userDao.getById(model.getAuth().getUid()) != null) {
            // Name, email address, and profile photo Url
            currentUser = userDao.getById(model.getAuth().getUid());
            String name = currentUser.getName(); //model.getUser().getDisplayName();
            String email = currentUser.getEmail(); //model.getUser().getEmail();
            //Uri photoUrl = model.getUser().getPhotoUrl();

               // Check if user's email is verified
            //boolean emailVerified = model.getUser().isEmailVerified();

            // The user's ID, unique to the Firebase project. Do NOT use this value to
            // authenticate with your backend server, if you have one. Use
            // FirebaseUser.getIdToken() instead.
            String uid = currentUser.getId(); // model.getUser().getUid();
            view.onClickRead();
            view.onLocalEnabled(name);
        } else if (model.getUser() != null){
            // Name, email address, and profile photo Url
            String name = model.getUser().getDisplayName();
            String email = model.getUser().getEmail();
            Uri photoUrl = model.getUser().getPhotoUrl();

            // Check if user's email is verified
            boolean emailVerified = model.getUser().isEmailVerified();


            // The user's ID, unique to the Firebase project. Do NOT use this value to
            // authenticate with your backend server, if you have one. Use
            // FirebaseUser.getIdToken() instead.
            String uid = model.getUser().getUid();
            view.onClickRead();
        }

    }

    /*void createAccount(String email, String password) {

        model.getAuth().createUserWithEmailAndPassword(email, password)
                .addOnCompleteListener(Objects.requireNonNull(view.getActivity()), task -> {
                    if (task.isSuccessful()) {
                        // Sign in success, update UI with the signed-in user's information
                        Toast.makeText(view.getActivity(), "User " + " with password",
                                Toast.LENGTH_LONG).show();
                        writeNewUser(model.getUser().getUid(), model.getUser().getEmail());
                        updateUI(model.getUser());
                    } else {
                        Toast.makeText(view.getActivity(), "Authentication failed.",
                                Toast.LENGTH_SHORT).show();
                        // If sign in fails, display a message to the user.
                        updateUI(null);
                    }

                });
    }*/

    void signIn(String email, String password) {
        /*if (userDao.getById(email) != null){
            view.onLocalEnabled();
        }*/
        model.getAuth().signInWithEmailAndPassword(email, password)
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        // Sign in success, update UI with the signed-in user's information
                        FirebaseUser user = model.getAuth().getCurrentUser();
                        // Toast.makeText(view.getActivity(), "User with password",

                        model.getUserData()
                                .addListenerForSingleValueEvent(new ValueEventListener() {
                                    @Override
                                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                        Iterable<DataSnapshot> dataChildren = dataSnapshot.getChildren();
                                        Iterator<DataSnapshot> iter = dataChildren.iterator();
                                        HashMap map = new HashMap();
                                        while (iter.hasNext()) {
                                            DataSnapshot ds = iter.next();
                                            String userFirebaseEmail = (String) ((HashMap) ds.getValue()).get("email");
                                            if(user.getEmail().equals(userFirebaseEmail)){
                                                String name = ((HashMap) ds.getValue()).get("name").toString();
                                                String id = ((HashMap) ds.getValue()).get("id").toString();
                                                String phone = ((HashMap) ds.getValue()).get("phone").toString();
                                                User curUser = new User(email, id, name, phone);
                                                userDao.insert(curUser);
                                                Log.d("Local_DB", "signIn with Network");
                                                userDao.update(curUser);
                                            }
                                        }
                                        updateUI(user);
                                    }

                                    @Override
                                    public void onCancelled(@NonNull DatabaseError databaseError) {

                                    }
                                });

                        //Toast.LENGTH_LONG).show();

                    } else {
                        // If sign in fails, display a message to the user.
                        Log.d("Local_DB", "signIn not completed");
                        updateUI(null);
                    }

                });
    }

    private void updateUI(FirebaseUser user) {
        if (user != null) {
            view.userValid(user);
        } else view.userNotValid();
    }

    private void updateUIForLocalDB(User user) {
        if (user != null) {
            view.userValidForLocalDB();
        } else view.userNotValid();
    }

    void updateUIFromPresenter() {
        updateUI(model.getUser());
    }




    /*
    Objects.requireNonNull(view.getActivity()), task -> {
                    if (task.isSuccessful()) {
                        // Sign in success, update UI with the signed-in user's information
                        FirebaseUser user = model.getAuth().getCurrentUser();
                        Toast.makeText(view.getActivity(), "User with password",
                                Toast.LENGTH_LONG).show();
                        updateUI(user);
                    } else {
                        // If sign in fails, display a message to the user.
                        Toast.makeText(view.getActivity(), "Authentication failed.",
                                Toast.LENGTH_SHORT).show();
                        updateUI(null);
                    }
                }
     */

}

File: app/src/main/java/ru/filchacov/billsplittest/authMVP/AuthPresenter.kt
package ru.filchacov.billsplittest.authMVP

import android.util.Log
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseUser
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import ru.filchacov.billsplittest.App
import ru.filchacov.billsplittest.ModelDB
import ru.filchacov.billsplittest.Singleton
import java.util.*

internal class AuthPresenter(private val view: AuthInterface) {
    private val model = ModelDB()
    private val userDB = App.getInstance().database
    private val userDao = userDB.userDao
    private lateinit var currentUser: FirebaseUser
    private val usersBillsDao = userDB.usersBillsDao
    private val singleton = App.getInstance().singleton
    private val billDao = userDB.billOfUserDao
    private val friendsIsChooseDao = userDB.friendsIsChooseDao
    private val friendsBillListDao = userDB.friendsBillListDao
    private val itemFromBillDao = userDB.itemFromBillDao
    private val savedFriendsDao = userDB.savedFriendsDao
    private val listDateTime: ArrayList<*> = ArrayList<String>()
    fun init() {
        model.authReference
        if (model.user != null) {
            view.onClickRead()
        }
    }

    fun signIn(email: String?, password: String?) {
        model.auth.signInWithEmailAndPassword(email!!, password!!)
                .addOnCompleteListener { task: Task<AuthResult?> ->
                    if (task.isSuccessful) {
                        val user = model.auth.currentUser
                        usersBillsDao.delete()
                        /* model.getUserData()
                                .addListenerForSingleValueEvent(new ValueEventListener() {
                                    @Override
                                    public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                                        Iterable<DataSnapshot> dataChildren = dataSnapshot.getChildren();
                                        for (DataSnapshot ds : dataChildren) {
                                            String userFirebaseEmail = (String) ((HashMap) ds.getValue()).get("email");
                                            assert user != null;
                                            if (user.getEmail().equals(userFirebaseEmail)) {
                                                String name = Objects.requireNonNull(((HashMap) ds.getValue()).get("name")).toString();
                                                String uid;
                                                if (((HashMap) ds.getValue()).get("userUid") != null) {
                                                    uid = ((HashMap) ds.getValue()).get("userUid").toString();
                                                } else {
                                                    uid = ((HashMap) ds.getValue()).get("id").toString();
                                                }
                                                String billUuid;
                                                HashMap friends = (HashMap) ((HashMap) ds.getValue()).get("friends");
                                                if (friends != null) {
                                                    for (Map.Entry<String, HashMap> friendsItem : (Iterable<Map.Entry<String, HashMap>>) friends.entrySet()) {
                                                        billUuid = friendsItem.getKey();
                                                        UsersBills usersBills = new UsersBills(billUuid);
                                                        usersBillsDao.insert(usersBills);
                                                        UUID friendUuid = UUID.randomUUID();
                                                        BillOfUser billOfUser = new BillOfUser(billUuid, friendUuid.toString());
                                                        billDao.insert(billOfUser);
                                                        listDateTime.add(billUuid);
                                                        HashMap billMap = friendsItem.getValue();
                                                        for (Object o : billMap.entrySet()) {
                                                            Map.Entry billMapItem = (Map.Entry) o;
                                                            if (!billMapItem.getKey().equals("savedFriends")) {
                                                                UUID chooseUUID = UUID.randomUUID();
                                                                FriendsIsChoose friendsIsChoose = new FriendsIsChoose(friendUuid.toString(), (String) billMapItem.getKey(), chooseUUID.toString());
                                                                friendsIsChooseDao.insert(friendsIsChoose);
                                                                ArrayList chooseList = (ArrayList) billMapItem.getValue();
                                                                for (int i = 0; i < chooseList.size(); i++) {
                                                                    HashMap friendBillItem = (HashMap) chooseList.get(i);
                                                                    UUID itemUUID = UUID.randomUUID();
                                                                    FriendsBillList friendsBillList =
                                                                            new FriendsBillList(
                                                                                    chooseUUID.toString(),
                                                                                    Integer.parseInt(friendBillItem.get("amount").toString()),
                                                                                    itemUUID.toString());
                                                                    friendsBillListDao.insert(friendsBillList);
                                                                    HashMap itemFromBillMap = (HashMap) friendBillItem.get("item");
                                                                    ItemFromBill itemFromBill = new ItemFromBill(itemUUID.toString(), (String) itemFromBillMap.get("name"), Integer.parseInt(itemFromBillMap.get("price").toString()), Integer.parseInt(itemFromBillMap.get("quantity").toString()), Integer.parseInt(itemFromBillMap.get("sum").toString()));
                                                                    itemFromBillDao.insert(itemFromBill);
                                                                }
                                                            } else {
                                                                HashMap savedFriend = (HashMap) billMapItem.getValue();
                                                                for (Object friend : savedFriend.entrySet()) {
                                                                    Map.Entry friendItem = (Map.Entry) friend;
                                                                    HashMap friendMap = (HashMap) friendItem.getValue();
                                                                    SavedFriends savedFriends = new SavedFriends(friendUuid.toString(), (Boolean) friendMap.get("isSelected"), friendItem.getKey().toString(), friendMap.get("mText").toString(), (friendMap.get("map") != null) ? Integer.parseInt(friendMap.get("sum").toString()) : 0);
                                                                    savedFriendsDao.insert(savedFriends);
                                                                }
                                                                billMapItem.getValue();
                                                            }
                                                        }

                                                    }
                                                }

                                                String phone = Objects.requireNonNull(((HashMap) ds.getValue()).get("phone")).toString();
                                                User curUser = new User(email, uid, name, phone);
                                                List<User> list = userDao.getAll();
                                                userDao.insert(curUser);
                                                Log.d("Local_DB", "signIn with Network");
                                                userDao.update(curUser);
                                                updateUI(user);
                                            }
                                        }
                                        view.btnEnable(true);
                                    }

                                    @Override
                                    public void onCancelled(@NonNull DatabaseError databaseError) {

                                    }
                                });*/singleton.someMethod(user, email)
                        CoroutineScope(Dispatchers.IO).launch {
                            var b = true
                            while (b) {
                                if (singleton.state == Singleton.COMPlETE) {
                                    withContext(Dispatchers.Main) {
                                        view.btnEnable(true)
                                        updateUI(user)
                                    }
                                    b = false
                                }
                            }
                        }
                        //Toast.LENGTH_LONG).show();
                    } else {
                        // If sign in fails, display a message to the user.
                        Log.d("Local_DB", "signIn not completed")
                        updateUI(null)
                        view.btnEnable(true)
                    }
                }
    }

    private fun updateUI(user: FirebaseUser?) {
        if (user != null) {
            view.userValid(user)
        } else view.userNotValid()
    }

    fun updateUIFromPresenter() {
        updateUI(model.user)
    }

}